lib/shared/s3-packages.js

```diff
@@ -0,0 +1,181 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const _ = require('lodash');
+const semver = require('semver');
+const Bluebird = require('bluebird');
+const request = Bluebird.promisifyAll(require('request'));
+const xml = Bluebird.promisifyAll(require('xml2js'));
+
+/**
+ * @summary Etcher production releases S3 bucket URL
+ * @constant
+ * @public
+ * @type {String}
+ */
+exports.BUCKET_URL_PRODUCTION = 'https://resin-production-downloads.s3.amazonaws.com';
+
+/**
+ * @summary Etcher snapshot releases S3 bucket URL
+ * @constant
+ * @public
+ * @type {String}
+ */
+exports.BUCKET_URL_SNAPSHOT = 'https://resin-nightly-downloads.s3.amazonaws.com';
+
+/**
+ * @summary Etcher releases S3 package name
+ * @constant
+ * @private
+ * @type {String}
+ */
+const S3_PACKAGE_NAME = 'etcher';
+
+/**
+ * @summary Number of packages per Etcher version
+ * @constant
+ * @private
+ * @type {Number}
+ */
+const NUMBER_OF_PACKAGES = 8;
+
+/**
+ * @summary Get the correct S3 bucket url from a version
+ * @function
+ * @public
+ *
+ * @param {String} version - application version
+ * @returns {String} S3 bucket url
+ *
+ * @example
+ * const version = require('../../package.json');
+ * const bucketUrl = s3Packages.getBucketUrlFromVersion(version);
+ * console.log(bucketUrl);
+ */
+exports.getBucketUrlFromVersion = (version) => {
+  const GIT_HASH_REGEX = /^[0-9a-f]{7,40}$/;
+
+  if (GIT_HASH_REGEX.test(_.first(semver.parse(version).build))) {
+    return exports.BUCKET_URL_SNAPSHOT;
+  }
+
+  return exports.BUCKET_URL_PRODUCTION;
+};
+
+/**
+ * @summary Get all remote versions from an S3 bucket
+ * @function
+ * @private
+ *
+ * @description
+ * We memoize based on the assumption that the received latest version
+ * number will not increase while the application is running.
+ *
+ * @param {String} bucketUrl - s3 bucket url
+ * @fulfil {String[]} - remote versions
+ * @returns {Promise}
+ *
+ * @example
+ * s3Packages.getRemoteVersions(s3Packages.BUCKET_URL_PRODUCTION).then((versions) => {
+ *   _.each(versions, (version) => {
+ *     console.log(version);
+ *   });
+ * });
+ */
+exports.getRemoteVersions = _.memoize((bucketUrl) => {
+
+  /* eslint-disable lodash/prefer-lodash-method */
+
+  return request.getAsync(bucketUrl)
+
+  /* eslint-enable lodash/prefer-lodash-method */
+
+    .get('body')
+    .then(xml.parseStringAsync)
+    .get('ListBucketResult')
+    .get('Contents')
+    .reduce((accumulator, entry) => {
+      const [ name, version ] = _.split(_.first(entry.Key), '/');
```

Comment:
I prefer to have them separate, given that this PR is already way too big. Hopefully we're close to merge this, so I can continue focusing on the other one.